docs/DafnyRef/integration- rust/IntegrationRust.md

```diff
@@ -71,42 +71,114 @@ For maximum flexibility, Dafny follows these rules to encode types in Rust:
+| `a[0]` if `a: array<T>`       | `rd!(a)[0]`             (*) |
+| `a[1] := 2` if  `a: array<T>` | `md!(a)[1] = 2`         (*) |
+
+
+## Encoding of classes and arrays with the option --raw-pointers
+The option `--raw-pointers` ensures that classes only have 8 bytes pointers,
+and array only have 8 bytes pointers + information about the length,
+and point to raw memory. Therefore, pointers have copy semantics and
+no call to any clone method needs to be made.
+However, since pointer comparison in Rust normally compares vtables,
+which is brittle, Dafny use its own pointer types which are wrappers:
+```rs
+pub struct Ptr<T: ?Sized>(pub Option<NonNull<UnsafeCell<T>>>);
+```
 
-## With the option --raw-pointers=true
-https://doc.rust-lang.org/book/appendix-07-nightly-rust.html
 |-------------------------------|-----------------------------|
 |  Dafny type                   |   Rust type                 |
 |-------------------------------|-----------------------------|
-| `C`, `C?` (for class, iterator C) | `*mut C`                |
-| (trait) `T`                   | (trait) `*mut dyn T`        |
-| `array<T>`                    | `*mut [T]`                  |
-| `array2<T>`                   | `*mut array2<T>`            |
+| `C`, `C?` (for class, iterator C) | `Ptr<C>`            (*) |
+| (trait) `T`                   | (trait) `Ptr<dyn T>`    (*) |
+| `array<T>`                    | `Ptr<[T]>`              (*) |
+| `array2<T>`                   | `Ptr<array2<T>>`    (*) (*) |
+| `c.x` if `c: C`               | `read!(c).x`            (*) |
+| `c.x := 2` if `c: C`          | `modify!(c).x = 2`      (*) |
+| `a[0]` if `a: array<T>`       | `read!(a)[0]`           (*) |
+| `a[1] := 2` if  `a: array<T>` | `modify!(a)[1] = 2`     (*) |
 
 (*) Defined in the Dafny runtime
 
+# Externs
+
+You can provide additional `*.rs` files to `dafny translate`, `dafny build` and even `dafny run` (via the `--input` option)
+All the elements imported from the Rust files are going to be available from the generated code via the module `_dafny_externs` that imports everything.
+
+The best way to see what you have to implement as an extern is to compile your code without it. You'll see that it will be necessary for the module `_dafny_extern` to use open a certain module, which you'd put in a file `external.rs` and pass it for compilation.
+
+When using simple 1-argument externs, the compilation follows Dafny conventions:
+* Static methods are static instance methods of a `pub struct _default {}` which is defined in the module itself, or must be defined externally if all static methods are labelled as externs.
+* Methods or functions with an extern attribute with 2 arguments are interpreted as follow: The first argument is a module name, and all the "." are rewritten to "::" to follow Rust's syntax. The second argument is the name of a static `pub fn` of that module.
+
+Let's assume you provide an additional input file `external.rs` with the following content:
```

Comment:
This feature does seem useful and a logical way to support passing in extra target language files as the other backends do. It is definitely limiting to not support nested trees of Rust source files though, especially if you're trying to just add a bit of Dafny code to an existing codebase. It is worth thinking about how to support that use case better in the future, if we continue to assume that the output from Dafny has to be the root file for the crate (and breaking that assumption might be the solution).

Reply:
It's a good call.
Nested trees of Rust source files are a bit more complicated than folder structures. This is because normally Rust looks out for files, or folders with a "mod.rs" inside them, recursively.

We could think of a future Rust-specific CLI option like `--root:false` which could make it possible to emit a non-root Rust file, provided a name different than `lib` in `--output`. That way, users would have full control over the root files and which modules to include.